ddocs/medic/views/tasks_by_state/map.js

```diff
@@ -0,0 +1,6 @@
+function(doc) {
+  if (doc.type === 'task') {
+    var emission = doc.emission || {};
+    emit(doc.state, { dueDate: emission.dueDate, startDate: emission.startDate, endDate: emission.endDate });
+  }
+}
```

Comment:
This view is very generic (eg: it emits all tasks whether or not they are in a terminal state). I'm in two minds about this. 

If we only emitted the ones in the terminal state then querying it would be easier because you could emit the `endDate` as the key and then include that as the `endKey` and then you could just purge everything couchdb returns rather than having to manually check the value against a date.

However the way it's written now could be useful in future if we want to query for tasks in other states.

I'm leaning towards only emitting terminal tasks right now - if we need to change it later to make it more generic then we can do that and the index will be rebuilt and it's not the end of the world. The likelihood of us getting the view definition right before we know the use case is very low anyway.

eg...

```js
function(doc) {
  if (doc.type === 'task' && doc.emission && doc.emission.endDate) {
    var isTerminalState = ['Cancelled', 'Completed', 'Failed'].indexOf(doc.state) >= 0;
    if (isTerminalState) {
      emit(doc.emission.endDate);
    }
  }
}
```

Reply:
I've considered this approach. However there is one problem with it: 

The way purge works (and I now realize that the current approach doesn't 100% respect this rule) is that we would grab all docs, check them against the purge rules and purge the ones that match and unpurge the ones that don't match. 
Which means that we'd need to grab all task documents and check their state and emission endDate. 

With querying a "fixed" view, we just get the tasks that should be purged now, but never get tasks that were purged and now should be unpurged. 

I don't know whether we should expect tasks that were in terminal state to be ever updated. Should we?

Reply:
I think there's some code in webapp to disallow moving out of a terminal state. I'm 99% confident that we don't need to handle this case...

Reply:
Done!